Alerting: Add notification policy provisioning file export

JacobsonMT opened the pull request:
**What is this feature?**

This feature implements the provisioning API export endpoints for exporting notification policies as well as adds an export button to the notification policy tree in the UI.

`GET /api/v1/provisioning/policies/export`

Similar to the rule endpoints, it supports both json and yaml response types via Accept header as well as a `format` query parameter. Query param `download=true/false` that will set Content-Disposition to recommend initiating a download or inline display.

NOTE: Content-Type will be text/yaml for inline display and application/yaml for downloads. This is because the default behaviour in some browsers (at least Chrome/Firefox) is to download application/yaml regardless of Content-Disposition

**Why do we need this feature?**

This feature is intended to facilitate provisioning file creation for notification policies.

**Which issue(s) does this PR fix?**:

Partial #70004

**Special notes for your reviewer:**

OpenAPI Spec: https://editor.swagger.io/?url=https://raw.githubusercontent.com/grafana/grafana/jacobsonmt/provisioning-export-notification-policy/pkg/services/ngalert/api/tooling/post.json

UI export menu item available on root policy only:
![Screenshot from 2023-06-13 09-33-00](https://github.com/grafana/grafana/assets/8484471/d143f32c-bb60-4557-b84c-f0b6443f18a1)

<details>
<summary>Example Output:</summary>

```yaml
apiVersion: 1
policies:
    - orgId: 2
      receiver: grafana-default-email
      group_by:
        - grafana_folder
        - alertname
      routes:
        - receiver: slack
          object_matchers:
            - - team
              - =
              - teamA
          continue: true
          routes:
            - receiver: oncall
              object_matchers:
                - - severity
                  - =
                  - critical
              group_wait: 30s
              group_interval: 5m
        - receiver: teams
          object_matchers:
            - - team
              - =
              - teamB
          continue: true
          routes:
            - receiver: oncall
              object_matchers:
                - - severity
                  - =
                  - critical
              group_wait: 30s
              group_interval: 10m
        - receiver: email
          object_matchers:
            - - team
              - =
              - teamC
          continue: true
          routes:
            - receiver: oncall
              object_matchers:
                - - severity
                  - =
                  - critical
              group_wait: 30s
              group_interval: 20m
```

</details>

JacobsonMT committed:
Alerting: Add notification policy provisioning file export

- Add provisioning API endpoint for exporting notification policies.
- Add option in notification policy view ellipsis dropdown for exporting.
- Update various provisioning documentation.

brendamuir reviewed (commented):
docs updates look good to me

alexweav commented on pkg/services/ngalert/api/tooling/definitions/provisioning.go:
```diff
@@ -0,0 +1,9 @@
+
+// AlertingFileExport is the full provisioned file export.
+// swagger:model
+type AlertingFileExport struct {
+	APIVersion int64                      `json:"apiVersion" yaml:"apiVersion"`
+	Groups     []AlertRuleGroupExport     `json:"groups,omitempty" yaml:"groups,omitempty"`
+	Policies   []NotificationPolicyExport `json:"policies,omitempty" yaml:"policies,omitempty"`
+}
```
Alertmanager (and therefore provisioning) will reject policies which reference contact points that aren't defined. So I think since this file doesn't contain any contact points, the files it produces are not importable. Should we implement some form of export for contact points first? Same goes for mute timings.

JacobsonMT replied:
Correct, any referenced contact points and mute timings referenced need to already exist otherwise if you try to import the exported policy it will fail. This makes it less useful than it could be with the other export endpoints but still functional.

The rest of the exports [are on the way](https://github.com/grafana/grafana/issues/70004) and I'll be working on Contact Points and Mute Timings next.

gillesdemey commented on pkg/services/ngalert/api/authorization_test.go:
```diff
@@ -49,7 +49,7 @@ func TestAuthorize(t *testing.T) {
 		paths[p] = methods
 	}
-	require.Len(t, paths, 48)
+	require.Len(t, paths, 49)
```
Magic number? Do we want to extract in to a separate var and add code comments?

JacobsonMT replied:
It's a strange test, in my experience it's here basically to remind you to add an entry in authorization.go if you've forgotten. It's reminded me a couple times, so I think it's doing something at least.

yuri-tceretian replied:
It makes sure that all paths are covered by authorization :)

yuri-tceretian reviewed (approved):
Backend code LGTM. However, I think the export button in UI is in place that is hard to find, in my opinion. I wonder if it would be much visible if we put a button on the page instead of a drop down menu for root policy that is behind a button that is hard to spot.

gillesdemey reviewed (approved):
UI changes LGTM

JacobsonMT committed:
Remove unintentionally added param endpoints links

JacobsonMT committed:
Regenerate api

grobinson-grafana reviewed (approved):
Test and it works, LGTM!

JacobsonMT committed:
Fix docs from rebase

JacobsonMT commented:
Rebased and fixed the conflicts. FYI @gillesdemey, needed to make some changes to accommodate the new `ConditionalWrap`s and tooltips. Also, added a check to make sure we only display this for the embedded grafana alertmanager.

JacobsonMT committed:
Fix linting and increment auth api test
